libpretixprint/src/main/java/eu/pretix/libpretixprint/templating/Layout.java

```diff
@@ -177,27 +196,133 @@ private void drawTextarea(JSONObject data, String text, PdfContentByte cb) throw
             alignment = Element.ALIGN_RIGHT;
         }
         para.setAlignment(alignment);
+        para.setKeepTogether(true);
 
         float lineheight;
-        if (data.has("lineheight")) {
-            lineheight = (float) (data.getDouble("lineheight") * 1.15);
+        if (data.has("lineheight") || !isLegacy) {
+            lineheight = (float) (data.optDouble("lineheight", 1.0) * 1.15);
         } else {
             lineheight = 1;
         }
-        para.setLeading(lineheight * fontsize);
+        para.setLeading((float) (lineheight * fontsize));
+
+        float adheight = (float) (Math.max(fontsize, baseFont.getAscentPoint(text, (float) fontsize) - baseFont.getDescentPoint(text + ",;gyjq", (float) fontsize)) + 0.0001);
+        float adheightPlusLH = adheight;
+        if (lineheight != 1) {
+            adheightPlusLH += (lineheight - 1.0) * fontsize;
+        }
+        return new ParagraphResult(para, fontsize, alignment, lineheight, adheightPlusLH, font);
+    }
+
+    private void drawTextcontainer(JSONObject data, String text, PdfContentByte cb) throws IOException, DocumentException, JSONException {
+        float width = millimetersToPoints((float) (data.getDouble("width")));
+        float height = millimetersToPoints((float) (data.getDouble("height")));
+        float widthextension = 0;
+        boolean autoresize = data.optBoolean("autoresize", false);
+        double fontSize = data.getDouble("fontsize");
+
+        ParagraphResult para;
+        double realHeight, realWidth;
+        while (true) {
+            para = getParagraph(data, text, false, fontSize);
+
+            ColumnText ct = new ColumnText(cb);
+            ct.addElement(para.p);
+            ct.setSimpleColumn(
+                    0,
+                    0,
+                    width + widthextension,
+                   3000000,
+                    (float) para.fontSize,
+                    para.alignment
+            );
+            ct.go(true);
+            realHeight = 3000000 - ct.getYLine();
+            realWidth = ct.getFilledWidth();
+
+            var smallEnough = (realHeight <= height && realWidth <= width);
+            if (!data.optBoolean("splitlongwords", true) && linesRequiredHardBreak(text, para.font, width + widthextension)) {
```

Comment:
See comment above, I think it reads better (and semantically seems more right) to call the method `linesRequireHardBreak`